### PR TITLE
Bucket policy to allow access for the root user

### LINKED ISSUE #770: terragrunt s3 bucket creation fails when using root credentials
When using the root credentials from a newly created AWS account, the S3 bucket is created but the root user doesn't have permissions to the newly created bucket.

The issue is probably that we don't add a policy that grants the root-user access per the [troubleshooting guide].(https://docs.aws.amazon.com/IAM/latest/UserGuide/troubleshoot_iam-s3.html)

Running the command as an IAM User with AdministratorAccess yields the expected result.

```
$ terragrunt init                                                                                                                                                                                  
[terragrunt] [/Users/matt/code/src/github.com/mcalhoun/infrastructure/live/root/_global/organization] 2019/07/01 17:03:59 Running command: terraform --version
[terragrunt] 2019/07/01 17:03:59 Reading Terragrunt config file at /Users/matt/code/src/github.com/mcalhoun/infrastructure/live/root/_global/organization/terragrunt.hcl
[terragrunt] 2019/07/01 17:03:59 WARNING: no double-slash (//) found in source URL /Users/matt/code/src/github.com/mcalhoun/infrastructure/modules/organization. Relative paths in downloaded Terraform code may not work.
[terragrunt] 2019/07/01 17:03:59 Downloading Terraform configurations from file:///Users/matt/code/src/github.com/mcalhoun/infrastructure/modules/organization into /Users/matt/code/src/github.com/mcalhoun/infrastructure/live/root/_global/organization/.terragrunt-cache/WnMUGlp2BJCICBcDh_ddHo4u-xQ/Q0y8bTwcIJYLO1ORZJW6dyJl7OE
[terragrunt] 2019/07/01 17:03:59 Copying files from /Users/matt/code/src/github.com/mcalhoun/infrastructure/live/root/_global/organization into /Users/matt/code/src/github.com/mcalhoun/infrastructure/live/root/_global/organization/.terragrunt-cache/WnMUGlp2BJCICBcDh_ddHo4u-xQ/Q0y8bTwcIJYLO1ORZJW6dyJl7OE
[terragrunt] 2019/07/01 17:03:59 Setting working directory to /Users/matt/code/src/github.com/mcalhoun/infrastructure/live/root/_global/organization/.terragrunt-cache/WnMUGlp2BJCICBcDh_ddHo4u-xQ/Q0y8bTwcIJYLO1ORZJW6dyJl7OE
[terragrunt] 2019/07/01 17:03:59 Initializing remote state for the s3 backend
[terragrunt] 2019/07/01 17:03:59 [terragrunt]  Remote state S3 bucket matt-calhoun-root-state does not exist or you don't have permissions to access it. Would you like Terragrunt to create it? (y/n)
y
[terragrunt] 2019/07/01 17:04:02 Create S3 bucket matt-calhoun-root-state
[terragrunt] 2019/07/01 17:04:02 Creating S3 bucket matt-calhoun-root-state
[terragrunt] 2019/07/01 17:04:02 Waiting for bucket matt-calhoun-root-state to be created
**[terragrunt] 2019/07/01 17:04:02 S3 bucket matt-calhoun-root-state created.**
[terragrunt] 2019/07/01 17:04:02 No tags specified for bucket matt-calhoun-root-state.
[terragrunt] 2019/07/01 17:04:02 Enabling versioning on S3 bucket matt-calhoun-root-state
[terragrunt] 2019/07/01 17:04:02 Enabling bucket-wide SSE on AWS S3 bucket matt-calhoun-root-state
[terragrunt] 2019/07/01 17:04:03 Granting WRITE and READ_ACP permissions to S3 Log Delivery (http://acs.amazonaws.com/groups/s3/LogDelivery) for bucket matt-calhoun-root-state. This is required for access logging.
[terragrunt] 2019/07/01 17:04:03 Waiting for ACL bucket matt-calhoun-root-state to have the updated ACL for access logging.
[terragrunt] 2019/07/01 17:04:03 Bucket matt-calhoun-root-state still does not have the ACL permissions for access logging. Will sleep for 5s and check again.
[terragrunt] 2019/07/01 17:04:08 Bucket matt-calhoun-root-state now has the proper ACL permissions for access logging!
[terragrunt] 2019/07/01 17:04:08 Enabling bucket-wide Access Logging on AWS S3 bucket "matt-calhoun-root-state" - using as TargetBucket "matt-calhoun-root-state"
[terragrunt] 2019/07/01 17:04:08 Lock table terraform-state-locks does not exist in DynamoDB. Will need to create it just this first time.
[terragrunt] 2019/07/01 17:04:08 Creating table terraform-state-locks in DynamoDB
[terragrunt] 2019/07/01 17:04:08 Table terraform-state-locks is not yet in active state. Will check again after 10s.
[terragrunt] 2019/07/01 17:04:18 Success! Table terraform-state-locks is now in active state.
[terragrunt] 2019/07/01 17:04:18 No tags for lock table given.
[terragrunt] 2019/07/01 17:04:18 Running command: terraform init -backend-config=dynamodb_table=terraform-state-locks -backend-config=encrypt=true -backend-config=key=_global/organization/terraform.tfstate -backend-config=region=us-east-1 -backend-config=bucket=matt-calhoun-root-state

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

Error: Failed to get existing workspaces: AccessDenied: Access Denied
	status code: 403, request id: 149E20E34744EA3A, host id: 39ECU1efjpQEfG+s/GrHm2hrrk3BU+y4q3LaqExrSYUcM5Bz/aMGvDqkoZUR1hCAuZZf
```

### PR DESCRIPTION
**Pull request by bwhaley**
After creating the remote state S3 bucket, we need to grant access to
the root user. Without this, the root user cannot use the remote state
bucket. This is [documented by AWS](https://docs.aws.amazon.com/IAM/latest/UserGuide/troubleshoot_iam-s3.html). This also adds a `skip_bucket_root_access` option to the remote state config in case this behavior is undesirable for some reason.

Fixes #770.

### CODE REVIEW

File: remote/remote_state_s3.go [hunk 1]
```diff
@@ -458,6 +468,62 @@ func isBucketAlreadyOwnedByYourError(err error) bool {
...
+		return "", err
+	}
+
+	identity, err := sts.New(session).GetCallerIdentity(nil)
+	if err != nil {
+		return "", errors.WithStackTrace(err)
+	}
+
+	return *identity.Account, nil
+}
```
[the remark this answers is not in the document]
Comment (bwhaley) [reply remote/remote_state_s3.go]: https://github.com/gruntwork-io/terragrunt/pull/978/commits/a4e647a201e0c10b61505504e2bc0ed29b9b26d8

File: remote/remote_state_s3.go [hunk 2]
```diff
@@ -458,6 +468,62 @@ func isBucketAlreadyOwnedByYourError(err error) bool {
...
+
+// Create the S3 bucket specified in the given config
+func EnableRootAccesstoS3Bucket(s3Client *s3.S3, config *RemoteStateConfigS3, terragruntOptions *options.TerragruntOptions) error {
```
[the remark this answers is not in the document]
Comment (bwhaley) [reply remote/remote_state_s3.go]: https://github.com/gruntwork-io/terragrunt/pull/978/commits/a4e647a201e0c10b61505504e2bc0ed29b9b26d8

**Comment by bwhaley** [comment]
Thanks for the review. Good suggestion on refactoring the AWS identity functions. I chose to pull them in to the `aws_helper` package instead. There is still a bit of redundancy since the wrappers in the `config` package have some opaque (to me) function signatures that don't lend themselves to a simple refactor - it seemed out of scope for this change. But happy to work on that too if I can get a little guidance on how those ctx wrappers work.

**Comment by yorinasub17** [comment]
> There is still a bit of redundancy since the wrappers in the config package have some opaque (to me) function signatures that don't lend themselves to a simple refactor - it seemed out of scope for this change

Yup those function signatures are necessary so that they can be available in the HCL parsing pipeline. We don't really want to touch those, so what you have done was exactly what I expected.

**Comment by bwhaley** [comment]
Tests are passing after I've merged upstream and fixed conflicts (docs related). I'll go ahead and merge now. Thanks for the feedback.

**Comment by bwhaley** [comment]
I had merged your changes in to my branch and it didn't drop anything. Looks like `skip_bucket_root_access` made it to the live site: https://terragrunt.gruntwork.io/use-cases/keep-your-remote-state-configuration-dry/

**Comment by brikis98** [comment]
OK, great, thx for checking

**Pull request merged**